### PR TITLE
feat(module): New VPN Client module

### PR DESCRIPTION
**Pull request by r351574nc3**
## what
* VPN Client Module for Terraform AWS

## why
* Allow configuration and provisioning of VPN infrastructure on AWS via manual (certificate)/federated (SSO) methods of authentication

## references
N/A

## todo
- [x] use security group module
- [ ] update terratests and confirm `/test all`
- [ ] address open comments

### CODE REVIEW

File: README.yaml [hunk 1]
```diff
@@ -73,15 +72,15 @@ usage: |-
 
   ```hcl
   module "example" {
-    source = "https://github.com/cloudposse/terraform-example-module.git?ref=master"
+    source = "https://github.com/cloudposse/terraform-aws-ec2-client-vpn?ref=master"
```
Comment (osterman) [inline README.yaml:75]: use registry notation

File: examples/complete/fixtures.us-east-2.tfvars [hunk 2]
```diff
@@ -8,3 +8,16 @@ stage = "test"
 
 name = "example"
 
+client_cidr = "0.0.0.0/24"
+
+aws_subnet_id = "sn-1234"
```
Comment (osterman) [inline examples/complete/fixtures.us-east-2.tfvars:13]: https://github.com/cloudposse/terraform-aws-eks-cluster/blob/master/examples/complete/main.tf#L35-L60

File: variables.tf [hunk 3]
```diff
@@ -1,4 +1,44 @@
...
+  description = "Retention in days for CloudWatch Log Group"
+}
+
+variable "internet_access_enabled" {
+  type        = bool
+  default     = true
+  description = <<-EOT
+Enables an authorization rule and route for the VPN to access the internet.
+Please note, you must allow ingress/egress to the internet (0.0.0.0/0) via the Subnet's security group.
+EOT
```
Comment (osterman) [inline variables.tf:38]:
```suggestion
  description = <<-EOT
    Enables an authorization rule and route for the VPN to access the internet.
    Please note, you must allow ingress/egress to the internet (0.0.0.0/0) via the Subnet's security group.
  EOT
```

File: main.tf [hunk 4]
```diff
@@ -1,13 +1,92 @@
...
+}
+
+resource "aws_cloudwatch_log_group" "vpn" {
+  name              = "/aws/vpn/${module.this.id}/logs"
+  retention_in_days = var.logs_retention
+  tags              = module.this.tags
+}
+
+resource "aws_cloudwatch_log_stream" "vpn" {
+  name           = "${module.this.id}-vpn-usage"
```
Comment (osterman) [inline main.tf:81]: we should use labels for this because even the `-` is a parameter of labels.

Comment (osterman) [reply main.tf:81]: you can add `vpn` and `usage` to the attributes

Comment (osterman) [reply main.tf:81]: https://github.com/cloudposse/terraform-null-label/blob/cf38625a5dde227db04c9cfedc1327d610229fec/examples/complete/label7.tf#L17

Comment (osterman) [reply main.tf:81]: I think `usage` doesn't add any extra disambiguation. It's already clear that it's a log events.

File: main.tf [hunk 5]
```diff
@@ -1,13 +1,92 @@
-resource "random_integer" "example" {
-  count = module.this.enabled ? 1 : 0
+provider "awsutils" {
+  region = "us-east-1"
+}
+
+module "tls_self_signed" {
+  source = "./modules/tls-self-signed"
```
Comment (osterman) [inline main.tf:6]: https://github.com/cloudposse/terraform-aws-ssm-tls-self-signed-cert

File: main.tf [hunk 6]
```diff
@@ -1,13 +1,92 @@
-resource "random_integer" "example" {
-  count = module.this.enabled ? 1 : 0
+provider "awsutils" {
+  region = "us-east-1"
```
Comment (osterman) [inline main.tf:2]: I think we typically use fixtures for this

File: main.tf [hunk 7]
```diff
@@ -1,13 +1,92 @@
...
+resource "aws_ec2_client_vpn_route" "internet_route" {
+  count = var.internet_access_enabled ? 1 : 0
+
+  client_vpn_endpoint_id = aws_ec2_client_vpn_endpoint.default.id
+  destination_cidr_block = "0.0.0.0/0"
+  target_vpc_subnet_id   = aws_ec2_client_vpn_network_association.default.subnet_id
+}
+
+resource "aws_cloudwatch_log_group" "vpn" {
+  name              = "/aws/vpn/${module.this.id}/logs"
```
Comment (osterman) [inline main.tf:75]:
```suggestion
  name              =  module.this.id
```

File: main.tf (hunk 4 above)
Comment (osterman) [reply main.tf:81]: Actually, lets use this module https://github.com/cloudposse/terraform-aws-cloudwatch-logs

File: main.tf (hunk 7 above)
Comment (osterman) [reply main.tf:75]: https://github.com/cloudposse/terraform-aws-cloudwatch-logs

File: main.tf [hunk 8]
```diff
@@ -1,13 +1,134 @@
...
+  certificate_chain = module.self_signed_cert_ca.certificate_pem
+}
+
+module "cloudwatch_log" {
+  source = "cloudposse/cloudwatch-logs/aws"
+
+  stream_names = var.stream_names
+
+  context = module.this.context
+}
```
Comment (osterman) [inline main.tf:79]:
```suggestion
module "cloudwatch_log" {
  source = "cloudposse/cloudwatch-logs/aws"

  stream_names = var.stream_names
  enabled = var.logging_enabled

  context = module.this.context
}
```

File: main.tf [hunk 9]
```diff
@@ -1,13 +1,134 @@
...
+
+  authentication_options {
+    type                       = "certificate-authentication"
+    root_certificate_chain_arn = aws_acm_certificate.root.arn
+  }
+
+  connection_log_options {
+    enabled               = var.logging_enabled
+    cloudwatch_log_group  = var.logging_enabled ? module.cloudwatch_log.log_group_name : null
+    cloudwatch_log_stream = var.logging_enabled ? element(var.stream_names, 0) : null
```
Comment (osterman) [inline main.tf:94]:
```suggestion
    cloudwatch_log_stream = var.logging_enabled ? var.logging_stream_name : null
```

File: main.tf [hunk 10]
```diff
@@ -1,13 +1,134 @@
...
+  description            = module.this.id
+  server_certificate_arn = aws_acm_certificate.server.arn
+  client_cidr_block      = var.client_cidr
+
+  authentication_options {
+    type                       = "certificate-authentication"
+    root_certificate_chain_arn = aws_acm_certificate.root.arn
+  }
+
+  connection_log_options {
```
Comment (osterman) [inline main.tf:91]: this should maybe be dynamic block

File: main.tf [hunk 11]
```diff
@@ -1,13 +1,134 @@
...
+  region = var.region
+}
+
+module "self_signed_cert_ca" {
+  source = "cloudposse/ssm-tls-self-signed-cert/aws"
+
+  name = "self-signed-cert-ca"
+
+  subject = {
+    common_name  = module.this.id
```
Comment (osterman) [inline main.tf:11]:
```suggestion
    common_name  = var.ca_common_name ? var.ca_common_name : module.this.id
```

File: main.tf [hunk 12]
```diff
@@ -1,13 +1,134 @@
...
+    cloudwatch_log_group  = var.logging_enabled ? module.cloudwatch_log.log_group_name : null
+    cloudwatch_log_stream = var.logging_enabled ? element(var.stream_names, 0) : null
   }
+
+  tags = module.this.tags
+}
+
+resource "aws_ec2_client_vpn_network_association" "default" {
+  client_vpn_endpoint_id = aws_ec2_client_vpn_endpoint.default.id
+  subnet_id              = var.aws_subnet_id
```
Comment (osterman) [inline main.tf:102]: We should be able to pass in a list of subnets

File: main.tf [hunk 13]
```diff
@@ -1,13 +1,134 @@
...
+}
+
+resource "aws_ec2_client_vpn_authorization_rule" "internal" {
+  client_vpn_endpoint_id = aws_ec2_client_vpn_endpoint.default.id
+  target_network_cidr    = var.aws_authorization_rule_target_cidr
```
Comment (osterman) [inline main.tf:107]: Support a list of target CIDRs

File: main.tf [hunk 14]
```diff
@@ -1,13 +1,134 @@
...
+  connection_log_options {
+    enabled               = var.logging_enabled
+    cloudwatch_log_group  = var.logging_enabled ? module.cloudwatch_log.log_group_name : null
+    cloudwatch_log_stream = var.logging_enabled ? element(var.stream_names, 0) : null
   }
+
+  tags = module.this.tags
+}
+
+resource "aws_ec2_client_vpn_network_association" "default" {
```
Comment (osterman) [inline main.tf:100]: This should use security groups too. Use our security groups module.

File: examples/complete/fixtures.us-east-2.tfvars [hunk 15]
```diff
@@ -8,3 +8,18 @@ stage = "test"
...
```
Comment (osterman) [inline examples/complete/fixtures.us-east-2.tfvars:13]:
```suggestion
```

File: docs/terraform.md [hunk 16]
```diff
@@ -3,57 +3,43 @@
...
 ## Outputs
 
 | Name | Description |
 |------|-------------|
-| <a name="output_example"></a> [example](#output\_example) | Example output |
-| <a name="output_id"></a> [id](#output\_id) | ID of the created example |
-| <a name="output_random"></a> [random](#output\_random) | Stable random number for this example |
+| example | Example output |
+| id | ID of the created example |
+| random | Stable random number for this example |
```
Comment (osterman) [inline docs/terraform.md:43]: This looks out of date

File: main.tf [hunk 17]
```diff
@@ -1,13 +1,203 @@
...
+  server_certificate_arn = aws_acm_certificate.server.arn
+  client_cidr_block      = var.client_cidr
+
+  authentication_options = local.mutual_enabled ? {
+    type                       = "certificate-authentication"
+    root_certificate_chain_arn = aws_acm_certificate.root.arn
+  } : {
+    type              = "federated-authentication"
+    saml_provider_arn = try(aws_iam_saml_provider.this[0].arn, var.saml_provider_arn)
+  }
```
Comment (osterman) [inline main.tf:114]: do something like....

```suggestion
  authentication_options = local.authentication_type == "certificate-authentication" ? {
    type                       = "certificate-authentication"
    root_certificate_chain_arn = aws_acm_certificate.root.arn
  } : {
    type              = "federated-authentication"
    saml_provider_arn = try(aws_iam_saml_provider.this[0].arn, var.saml_provider_arn)
  }
```

File: outputs.tf [hunk 18]
```diff
@@ -1,14 +1,9 @@
...
-}
-
-output "random" {
-  description = "Stable random number for this example"
-  value       = module.this.enabled ? join("", random_integer.example[*].result) : null
-}
+output "client_configuration" {
+  value       = awsutils_ec2_client_vpn_export_client_config.default.client_configuration
+  description = "VPN Client Configuration data."
+}
```
Comment (osterman) [inline outputs.tf:9]: Lets also add

arn - The ARN of the Client VPN endpoint.
dns_name - The DNS name to be used by clients when establishing their VPN session.
id - The ID of the Client VPN endpoint.

Comment (osterman) [reply outputs.tf:9]: Maybe add an optional DNS name

https://registry.terraform.io/modules/cloudposse/route53-cluster-hostname/aws/latest

File: variables.tf [hunk 19]
```diff
@@ -1,4 +1,137 @@
...
+  type = list(object({
+    destination_cidr_block = string
+    description            = string
+    target_vpc_subnet_id   = string
+  }))
+}
+
+variable "additional_security_groups" {
+  default     = []
+  description = "List of security groups to attach to the client vpn network associations"
```
Comment (osterman) [inline variables.tf:119]: I think we are not using our security group module

Comment (osterman) [reply variables.tf:119]: https://github.com/cloudposse/terraform-aws-security-group

Comment (osterman) [reply variables.tf:119]: https://github.com/cloudposse/terraform-aws-elasticache-redis/blob/master/main.tf#L17-L28

File: main.tf [hunk 20]
```diff
@@ -1,13 +1,203 @@
...
+  vpc_id      = var.vpc_id
+
+  ingress {
+    description = "Allow self access only by default"
+    from_port   = 0
+    protocol    = -1
+    self        = true
+    to_port     = 0
+  }
+}
```
Comment (osterman) [inline main.tf:162]: https://github.com/cloudposse/terraform-aws-elasticache-redis/blob/master/main.tf#L17-L28

File: README.yaml [hunk 21]
```diff
@@ -57,8 +57,7 @@ references:
 
 # Short description of this project
 description: |-
-  This is `terraform-example-module` project provides all the scaffolding for a typical well-built Cloud Posse module. It's a template repository you can
-  use when creating new repositories.
+  The `terraform-aws-ec2-client-vpn` project provides for ec2 client vpn infrastructure.
```
Comment (osterman) [inline README.yaml:60]:
```suggestion
  The `terraform-aws-ec2-client-vpn` project provides for ec2 client vpn infrastructure. AWS Client VPN is a managed client-based VPN service based on OpenVPN that enables you to securely access your AWS resources and resources in your on-premises network. With Client VPN, you can access your resources from any location using [any OpenVPN-based VPN client](https://docs.aws.amazon.com/vpn/latest/clientvpn-user/connect-aws-client-vpn-connect.html).


```

File: README.yaml [hunk 22]
```diff
@@ -36,9 +36,9 @@ badges:
...
 related:
-  - name: "terraform-null-label"
-    description: "Terraform module designed to generate consistent names and tags for resources. Use terraform-null-label to implement a strict naming convention."
-    url: "https://github.com/cloudposse/terraform-null-label"
+  - name: "terraform-aws-components"
+    description: "Repository collection of aws components."
+    url: "https://github.com/cloudposse/terraform-aws-components"
 
 # List any resources helpful for someone to get started. For example, link to the hashicorp documentation or AWS documentation.
 references:
```
Comment (osterman) [inline README.yaml:44]:
```suggestion
references:
  - name: OpenVPN Clients
    description: "Any OpenVPN client should be compatible with the AWS Client VPN."
     url: https://docs.aws.amazon.com/vpn/latest/clientvpn-user/connect-aws-client-vpn-connect.html
```

File: main.tf [hunk 23]
```diff
@@ -1,13 +1,180 @@
-resource "random_integer" "example" {
-  count = module.this.enabled ? 1 : 0
+provider "awsutils" {
+  region = var.region
+}
+
+locals {
+  enabled           = module.this.enabled
+  mutual_enabled    = var.authentication_type == "certificate-authentication"
+  federated_enabled = var.authentication_type == "federated_authentication"
```
Comment (osterman) [inline main.tf:8]: double check this

File: main.tf [hunk 24]
```diff
@@ -1,13 +1,180 @@
...
+      cidr_blocks = ["0.0.0.0/0"]
+      description = "Allow self access only by default"
+      self        = true
+    },
+  ]
+
+  vpc_id = module.vpc.vpc_id
+
+  context = module.this.context
+}
```
Comment (osterman) [inline main.tf:139]: This doesn't resemble how we use the security group module elsewhere. The objective security group module is ot expose an interface that makes it easier for security-conscious companies to fine-tune the rules. Thus we want to expose rules as variables.

Comment (r351574nc3) [reply main.tf]: In this case, it is the default security group for the VPC. We want the rules fixed such that it is a minimum requirement.

File: variables.tf [hunk 25]
```diff
@@ -1,4 +1,154 @@
...
+  description = "Optional SAML provider ARN. Must include this or `saml_metadata_document`"
+  type        = string
+
+  validation {
+    error_message = "Invalid SAML provider ARN."
+
+    condition = (
+      var.saml_provider_arn == null ||
+      try(length(regexall(
+        "^arn:aws:iam::(?P<account_id>\\d{12}):saml-provider/(?P<provider_name>[\\w+=,\\.@-]+)$",
```
Comment (osterman) [inline variables.tf:91]: We should ensure that our modules are gov/china cloud compatible. The partition will change and not always be `aws`
```suggestion
        "^arn:[^:]+:iam::(?P<account_id>\\d{12}):saml-provider/(?P<provider_name>[\\w+=,\\.@-]+)$",
```

File: main.tf [hunk 26]
```diff
@@ -1,13 +1,191 @@
...
+
+  client_vpn_endpoint_id = aws_ec2_client_vpn_endpoint.default.id
+  target_network_cidr    = "0.0.0.0/0"
+  authorize_all_groups   = true
+}
+
+module "vpn_security_group" {
+  source = "cloudposse/security-group/aws"
+
+  attributes = ["simple"]
```
Comment (osterman) [inline main.tf:134]:
```suggestion
```

**Pull request merged**